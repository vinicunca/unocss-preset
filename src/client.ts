export * from './clients/masonry-grid-polyfill';
