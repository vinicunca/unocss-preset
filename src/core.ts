import { toKebabCase } from '@vinicunca/js-utilities';

export interface CoreOptions {
  prefix?: string;
  colorKeys?: string[];
}

export const DEFAULT_PREFIX = 'vin-';

export class PresetCore {
  prefix: string;
  colorKeys: string[];

  constructor(options: CoreOptions) {
    this.prefix = options.prefix ?? '';
    this.colorKeys = options.colorKeys ?? [];
  }

  genVariable(key = '') {
    return `--${this.prefix}${key}`;
  }

  remapVariables({ tokens, key = '' }: { tokens: Record<string, string | number>; key?: string }) {
    return Object.keys(tokens).reduce<any>((prev, acc) => {
      const _key = `${this.genVariable(key)}${toKebabCase(acc)}`;
      prev[_key] = tokens[acc];
      return prev;
    }, {});
  }

  getColorKeys() {
    return Array.from(new Set(this.colorKeys));
  }
}
