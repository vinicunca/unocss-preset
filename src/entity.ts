import type { ButtonOptions } from './components';

export const DEFAULT_PREFIX = 'vin-';

export interface IPresetVinicunca {
  prefix?: string;

  components?: {
    button?: ButtonOptions;
  };
}
