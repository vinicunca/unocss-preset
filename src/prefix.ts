export interface PrefixOptions {
  prefix: string;
}

export class PresetPrefix {
  prefix: string;

  constructor(prefix: string) {
    this.prefix = prefix;
  }
}
