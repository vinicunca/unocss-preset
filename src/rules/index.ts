import type { Rule } from '@unocss/core';

import { button } from './button';

export const rules: Rule[] = [
  button,
].flat(1);
