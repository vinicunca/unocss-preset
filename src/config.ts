import type { IConfigVinicunca } from './entity';
import type { Preset, Rule, Shortcut, SourceCodeTransformer } from '@unocss/core';

import { Color } from './base/color';
import { Elevation } from './base/elevation';
import { Theme } from './themes';
import { Button } from './components';
import { DEFAULT_PREFIX } from './core';

export class VincuncaConfig {
  private rules: Rule[];
  private shortcuts: Shortcut[];
  private prefix: string;

  private theme: Theme;

  // base
  private color: Color;
  private elevation: Elevation;

  // components
  private button: Button;

  constructor(options: IConfigVinicunca) {
    const { prefix, components, theme } = options;

    this.prefix = prefix ?? DEFAULT_PREFIX;

    this.theme = new Theme({
      prefix: this.prefix,
      ...theme,
    });

    const colorKeys = this.theme.getColorKeys();

    // init base
    this.color = new Color({
      prefix: this.prefix,
      colorKeys,
    });
    this.elevation = new Elevation();

    // init components
    this.button = new Button({
      prefix: this.prefix,
      colorKeys,
      ...components?.button,
    });

    this.rules = this.defineRules();

    this.shortcuts = this.defineShortcuts();
  }

  getPreset(): Preset {
    return {
      name: 'unocss-preset-vinicunca',
      layers: {
        vinicunca: -2,
        preflight: -1,
        default: 0,
        variants: 1,
      },
      prefix: this.prefix,
      preflights: [
        this.theme.getPreflight(),
      ],
      rules: this.rules,
      shortcuts: this.shortcuts,
    };
  }

  getTransformer(): SourceCodeTransformer {
    return {
      name: 'vinicunca-transformers',
      enforce: 'pre',
      transform: (str) => {
        let content = str.toString();

        content = this.button.transformClass(content);

        str.overwrite(0, str.length(), content);
      },
    };
  }

  private defineRules(): Rule[] {
    return [
      this.theme.getRules(),
      this.color.getRules(),
      this.elevation.getRules(),
      this.button.getRules(),
    ].flat(1);
  }

  private defineShortcuts(): Shortcut[] {
    return [
      ...this.button.getShortcuts(),
    ];
  }
}

