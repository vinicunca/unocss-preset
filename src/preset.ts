import type { IPresetVinicunca } from './entity';
import type { Rule, Shortcut } from '@unocss/core';

import { DEFAULT_PREFIX } from './entity';
import { Button } from './components';
import { PresetPrefix } from './prefix';

export class PresetVinicunca extends PresetPrefix {
  options: IPresetVinicunca;
  rules: Rule[];
  shortcuts: Shortcut[];

  // components
  button: Button;

  constructor(options: IPresetVinicunca) {
    super(options.prefix ?? DEFAULT_PREFIX);

    this.options = options;

    // init components
    this.button = new Button({
      prefix: this.prefix,
      ...options.components?.button,
    });

    this.rules = this.defineRules();

    this.shortcuts = this.defineShortcuts();
  }

  private defineRules(): Rule[] {
    return [
      this.button.getRules(),
    ].flat(1);
  }

  private defineShortcuts(): Shortcut[] {
    return [
      ...this.button.getShortcuts(),
    ];
  }

  getPresetConfigs() {
    return {
      name: 'unocss-preset-vinicunca',
      layers: {
        vinicunca: -1,
        variants: 1,
      },
      prefix: this.prefix,
      rules: this.rules,
      shortcuts: this.shortcuts,
    };
  }
}

