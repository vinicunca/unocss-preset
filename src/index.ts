import type { Preset } from '@unocss/core';
import type { IPresetVinicunca } from './entity';

import { PresetVinicunca } from './preset';

/**
 * Vinicunca Preset
 *
 * @example
 * // unocss.config.ts
 * import { presetUno } from 'unocss'
 * import { presetVinicunca } from '@vinicunca/unocss-preset'
 *
 * export default defineConfig({
 *   presets: [
 *     presetUno(),
 *     presetVinicunca()
 *   ]
 * })
 */
export function presetVinicunca(options: IPresetVinicunca = {}): Preset {
  const preset = new PresetVinicunca(options);

  return preset.getPresetConfigs();
}
