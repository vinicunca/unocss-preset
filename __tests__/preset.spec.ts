import { createGenerator, presetUno } from 'unocss';
import { describe, expect, test } from 'vitest';

import { presetVinicunca } from '../src';

const vinicunca = createGenerator({
  presets: [
    presetUno({
      preflight: false,
    }),
    presetVinicunca({
      components: {
        button: {
          blackContrast: '#000',
          whiteContrast: '#fff',
          sizes: {
            'x-small': '[--vin-height:20px] text-tiny min-w-[36px] px-2',
            'small': '[--vin-height:28px] text-xs min-w-[50px] px-3',
            'large': '[--vin-height:44px] text-base min-w-[78px] px-5',
            'x-large': '[--vin-height:52px] text-lg min-w-[92px] px-6',
          },
        },
      },
    }),
  ],
});

const classes = [
  'mt-4',
  'vin-button',
  'vin-button--large',
  'vin-button-blue-500',
  'vin-button-outline-red-100',
];

describe('preset-vinicunca', () => {
  test('core', async () => {
    const { css } = await vinicunca.generate(classes.join(' '));
    expect(css).toMatchSnapshot();
  });
});
