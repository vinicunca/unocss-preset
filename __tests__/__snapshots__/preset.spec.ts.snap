// Vitest Snapshot v1

exports[`preset-vinicunca > core 1`] = `
"/* layer: vinicunca */
.vin-button{--vin-border-color:#fff;}
.vin-button-blue-500{--vin-border-color:#3b82f6;--un-bg-opacity:1;--vin-color:#fff;--vin-bg-color:rgba(59,130,246,var(--un-bg-opacity));--vin-hover-color:#fff;--vin-hover-bg-color:rgb(118, 168, 249);--vin-hover-border-color:rgb(118, 168, 249);--vin-active-bg-color:rgb(51, 108, 201);--vin-active-border-color:rgb(51, 108, 201);--vin-disable-color:#fff;--vin-disable-bg-color:rgb(157, 193, 251);--vin-disable-border-color:rgb(157, 193, 251);}
.vin-button-outline-red-100{--vin-border-color:#fee2e2;--vin-color:#fee2e2;--vin-hover-color:#fff;--vin-hover-bg-color:#fee2e2;--vin-hover-border-color:#fee2e2;--vin-active-color:#fff;--vin-active-bg-color:rgb(207, 185, 185);--vin-active-border-color:rgb(207, 185, 185);--vin-disable-color:rgb(255, 241, 241);--vin-disable-border-color:rgb(255, 249, 249);}
/* layer: default */
.mt-4{margin-top:1rem;}
/* layer: variants */
.vin-button--large{--vin-height:44px;min-width:78px;padding-left:1.25rem;padding-right:1.25rem;font-size:1rem;line-height:1.5rem;}"
`;
